javalib/src/main/scala/java/util/Formatter.scala

```diff
@@ -1014,6 +1045,182 @@ object Formatter {
+
+      roundAtPos(roundingPos = precision)
+    }
+
+    /** Returns a new Decimal instance with the same value, possibly rounded,
+     *  with the given scale.
+     *
+     *  If this is a zero value, the same value is returned (a zero value must
+     *  always have a 0 scale). Rounding may also cause the result to be a zero
+     *  value, in which case its scale must be 0 as well. Otherwise, the result
+     *  is non-zero and is guaranteed to have exactly the given new scale.
+     */
+    def setScale(newScale: Int): Decimal = {
+      val roundingPos = unscaledValue.length() + newScale - scale
+      val rounded = roundAtPos(roundingPos)
+      // Assert: rounded.isZero || rounded.scale <= newScale
+
+      if (rounded.isZero || rounded.scale == newScale)
+        rounded
+      else
+        new Decimal(negative, rounded.unscaledValue + strOfZeros(newScale - rounded.scale), newScale)
+    }
+
+    /** Rounds the number at the given position in its `unscaledValue`.
+     *
+     *  The `roundingPos` may be any integer value.
+     *
+     *  - If it is < 0, the result is always a zero value.
+     *  - If it is >= `unscaledValue.lenght()`, the result is always the same
+     *    value.
+     *  - Otherwise, the `unscaledValue` will be truncated at `roundingPos`,
+     *    and rounded up iff `unscaledValue.charAt(roundingPos) >= '5'`.
+     *
+     *  The value of `negative` is always preserved.
+     *
+     *  Unless the result is a zero value, the following guarantees apply:
+     *
+     *  - its scale is guaranteed to be at most
+     *    `scale - (unscaledValue.length() - roundingPos)`.
+     *  - its precision is guaranteed to be at most
+     *    `max(1, roundingPos)`.
+     */
+    private def roundAtPos(roundingPos: Int): Decimal = {
+      val digits = this.unscaledValue // local copy
+      val digitsLen = digits.length()
+
+      if (roundingPos < 0) {
+        Decimal.zero(negative)
+      } else if (roundingPos >= digitsLen) {
+        this // no rounding necessary
+      } else {
+        @inline def scaleAtPos(pos: Int): Int = scale - (digitsLen - pos)
+
+        if (digits.charAt(roundingPos) < '5') {
+          // Truncate at roundingPos
+          if (roundingPos == 0)
+            Decimal.zero(negative)
+          else
+            new Decimal(negative, digits.substring(0, roundingPos), scaleAtPos(roundingPos))
+        } else {
+          // Truncate and increment at roundingPos
+
+          // Find the position of the last non-9 digit in the truncated digits (can be -1)
+          var lastNonNinePos = roundingPos - 1
+          while (lastNonNinePos >= 0 && digits.charAt(lastNonNinePos) == '9')
+            lastNonNinePos -= 1
+
+          val newUnscaledValue =
+            if (lastNonNinePos < 0) "1"
+            else digits.substring(0, lastNonNinePos) + (digits.charAt(lastNonNinePos) + 1).toChar
+
+          val newScale = scaleAtPos(lastNonNinePos + 1)
```

Comment:
Hah, it's kind of cool that the underflow at `-1` is actually meaningful here :)